### PR TITLE
[MODINVSTOR-1218]. Implement Domain Event Production for Institution Create/Update/Delete

### PR DESCRIPTION
**Pull request by BKadirkhodjaev**
### Purpose

[[MODINVSTOR-1218]. Implement Domain Event Production for Institution Create/Update/Delete](<https://folio-org.atlassian.net/browse/MODINVSTOR-1218>)

### Approach

- Add Kafka topic
- Add event publisher
- Add integration tests
- Update existing classes and tests
- Update README.md & ModuleDescriptor-template.json

### Changes Checklist
- [ ] **API Changes**: Document any API paths, methods, request or response bodies changed, added, or removed.
- [ ] **Database Schema Changes**: Indicate any database schema changes and their impact. Confirm that migration scripts were created.
- [ ] **Interface Version Changes**: Indicate any changes to interface versions.
- [ ] **Interface Dependencies**: Document added or removed dependencies.
- [ ] **Permissions**: Document any changes to permissions.
- [ ] **Logging**: Confirm that logging is appropriately handled.
- [ ] **Unit Testing**: Confirm that changed classes were covered by unit tests.
- [x] **Integration Testing**: Confirm that changed logic was covered by integration tests.
- [x] **Manual Testing**: Confirm that changes were tested on local or dev environment.
- [x] **NEWS**: Confirm that the NEWS file is updated with relevant information about the changes made in this pull request.

### Related Issues
https://folio-org.atlassian.net/browse/MODINVSTOR-1181
https://folio-org.atlassian.net/browse/MODINVSTOR-1216

### CODE REVIEW

**Review by psmagin — CHANGES_REQUESTED** [review]
Please fix Sonar

File: src/test/java/org/folio/rest/impl/LocationUnitInstitutionIT.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    var message =
+      String.format("id value already exists in table locinstitution: %s", id);
+
+    var successfulRecord =
+      doPost(client, resourceUrl(), pojo2JsonObject(record)).onComplete(
+        verifyStatus(ctx, HTTP_CREATED));
+    var failedRecord =
+      doPost(client, resourceUrl(), pojo2JsonObject(record)).onComplete(
+        verifyStatus(ctx, HTTP_UNPROCESSABLE_ENTITY));
+
```
[the remark this answers is not in the document]
Comment (BKadirkhodjaev) [reply src/test/java/org/folio/rest/impl/LocationUnitInstitutionIT.java:116]: From my observation chaining two doPost() commands (with one being nested in the other), the test sometimes results in 201 being produced when the expected is 422. For this reason I went form Future.all() solution.

[the remark this answers is not in the document]
Comment (BKadirkhodjaev) [reply src/test/java/org/folio/rest/impl/LocationUnitInstitutionIT.java]: I've update the test and tried on repeat. It is reliable: 
image

File: src/main/java/org/folio/services/locationunit/InstitutionService.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  public Future<Response> create(Locinst institution) {
+    return PgUtil.post(INSTITUTION_TABLE, institution, okapiHeaders,
+        vertxContext, PostLocationUnitsInstitutionsResponse.class)
+      .onSuccess(response -> {
+        domainEventService.publishCreated().handle(response);
+
+        var headers = PostLocationUnitsInstitutionsResponse.headersFor201()
+          .withLocation(URL_PREFIX + response);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/services/locationunit/InstitutionService.java:68]: do not know who use this locations header, as for me it is ok to have the same approach as in similar PR to make code simpler and cleaner:
https://github.com/folio-org/mod-inventory-storage/pull/1029/files#diff-a504c82181f298bf1db3c82e991febfb96b01ed3ed5d5295ea46050e1a2b2fcfR54
But let's wait what @psmagin will suggest

**Review by SerhiiNosko — APPROVED** [review]

File: src/main/java/org/folio/services/locationunit/InstitutionService.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    if (Objects.isNull(institution.getId())) {
+      institution.setId(id);
+    }
+
+    return repository.getById(id)
+      .compose(oldInstitution ->
+        PgUtil.put(INSTITUTION_TABLE, institution, id,
+          okapiHeaders, vertxContext,
+          PutLocationUnitsInstitutionsByIdResponse.class)
+        .onSuccess(domainEventService.publishUpdated(institution)));
```
Comment (azizbekxm) [inline src/main/java/org/folio/services/locationunit/InstitutionService.java:83]: please check this place in action. How it looks. We shuold have first oldInstitution and the next new instution in kafka message. 
example
image

Comment (BKadirkhodjaev) [reply src/main/java/org/folio/services/locationunit/InstitutionService.java]: > I suppose instutition should be oldInstutition. please check this place in action. How it looks. We shuold have first oldInstitution and the next new instution in kafka message. example image

Done, fixed my update code, here are topic messages on update:

image

**Review by azizbekxm — APPROVED** [review]

File: src/main/java/org/folio/services/locationunit/InstitutionService.java (hunk 2 above)
Comment (psmagin) [reply src/main/java/org/folio/services/locationunit/InstitutionService.java]: I'm okay with not having it

**Review by psmagin — APPROVED** [review]

**Pull request merged**